mqtt/mqtt_client.go

```diff
@@ -0,0 +1,174 @@
+	mqtt.CRITICAL = log.New(os.Stdout, "[mqtt:CRITICAL]", 0)
+	//mqtt.WARN = log.New(os.Stdout, "[mqtt:warn]", 0)
+
+	mqttScheme := getEnv("MQTT_SCHEME", "tcp")
+	mqttPort := getEnv("MQTT_PORT", "1883")
+	mqttHost := getEnv("MQTT_HOSTNAME", "host.docker.internal")
+	mqttClientID, _ := randomClientName()
+
+	password, mqttPublicKey, err := signTimestamp()
+	if err != nil {
+		panic(err)
+	}
+
+	brokerUri := fmt.Sprintf("%s://%s:%s", mqttScheme, mqttHost, mqttPort)
+
+	opts = mqtt.NewClientOptions()
+	opts.AddBroker(brokerUri)
+	opts.SetClientID(mqttClientID)
+
+	opts.SetUsername(mqttPublicKey)
+	opts.SetPassword(password)
+	opts.SetCleanSession(false)
+	opts.SetOnConnectHandler(func(c mqtt.Client) { connectedToLocalCallback(c, mqttPublicKey) })
+	opts.SetDefaultPublishHandler(f)
+}
+
+func Client() mqtt.Client {
+	client := mqtt.NewClient(opts)
+	if token := client.Connect(); token.Wait() && token.Error() != nil {
+		fmt.Println("=> connection error:", token.Error())
+	}
+	return client
+}
+
+func connectedToLocalCallback(c mqtt.Client, pubkey string) {
+	c.Subscribe(pubkey+"/#", 0, func(c mqtt.Client, msg mqtt.Message) {
+		fmt.Printf("* [%s] %s\n", msg.Topic(), string(msg.Payload()))
+	})
+	pwd, pubkey, err := signTimestamp()
+	if err != nil {
+		log.Println("there was an error:", err)
+	}
+	c.Publish(pubkey+"/"+pwd, 0, false, "hi")
+}
+
+func signTimestamp() (string, string, error) {
+	if thePrivateKey == nil {
+		b := make([]byte, 32)
+		_, err := rand.Read(b)
+		if err != nil {
+			return "", "", err
+		}
+		thePrivateKey = b
+	}
+
+	p, pub := btcec.PrivKeyFromBytes(btcec.S256(), thePrivateKey)
+	pubHex := hex.EncodeToString(pub.SerializeCompressed())
+
+	t := time.Now().Unix()
+	timeBuf := make([]byte, 4)
+	binary.BigEndian.PutUint32(timeBuf, uint32(t))
+	sig, err := Sign(timeBuf, p)
+	if err != nil {
+		panic(err)
+	}
+
+	pwdBuf := append(timeBuf, sig...)
+	return base64.URLEncoding.EncodeToString(pwdBuf), pubHex, nil
+}
+
+// NodeSigner is an implementation of the MessageSigner interface backed by the
+// identity private key of running lnd node.
+type nodeSigner struct {
+	privateKey *btcec.PrivateKey
+}
+
+// NewNodeSigner creates a new instance of the NodeSigner backed by the target
+// private key.
+func newNodeSigner(key *btcec.PrivateKey) *nodeSigner {
+	priv := &btcec.PrivateKey{}
+	priv.Curve = btcec.S256()
+	priv.PublicKey.X = key.X
+	priv.PublicKey.Y = key.Y
+	priv.D = key.D
+	return &nodeSigner{
+		privateKey: priv,
+	}
+}
+
+// Sign ...
+func Sign(msg []byte, privKey *btcec.PrivateKey) ([]byte, error) {
+
+	if privKey == nil || msg == nil {
+		return nil, errors.New("bad")
+	}
+	msg = append(signedMsgPrefix, msg...)
+	digest := chainhash.DoubleHashB(msg)
+	sigBytes, err := btcec.SignCompact(btcec.S256(), privKey, digest, true)
+	if err != nil {
+		return nil, err
+	}
+	return sigBytes, nil
+}
+
+func Publish(c mqtt.Client, topic string, payload string, wait bool) {
+	token := c.Publish(topic, 0, false, payload)
```

Comment:
When I'm calling publish im getting this error 
https://github.com/stakwork/sphinx-relay/pull/342#issuecomment-1087933979

I'm not sure if its because of the `username` and `password` in `opt` but I'm not familiar enough with mqtt to know if that's the issue

Reply:
The username/password used to connect to MQTT is a random public/private key & a signed timestamp using that key ( as per conversions with @Evanfeenstra ).
 
Is this error coming from the MQTT server or `sphinx-tribes`?

Reply:
I think you need to re-work the mqtt Client. Check out db.go, there is a global object called DB, that has methods you can all to do things. The mqtt client should work like that... it connects once at the beginning (with the Init) call, and then you can call it any time to publish a msg. Right now, you are passing in a mqtt.Client to the Publish call, but that Client is not at all related to the `Init` function? In fact it looks like the `Init` function just makes a NewClientOptions struct but then doesn't actually do anything with it?

Reply:
done, however there is still an error on startup from the client:

```
Init MQTT Client...
[mqtt:error][client]   Failed to connect to a broker
=> connection error: Not Authorized
```

also, side note, this is the request that I've been using to hit the endpoint 
```
wget --no-check-certificate --quiet \
  --method POST \
  --timeout=0 \
  --header 'content-type: application/json' \
  --body-data '{
    "chat_uuid":"{{CHAT_UUID}}",
    "batch_contents":[
        {"mqtt_topic":"{{MQTT_TOPPIC}}",
        "data":"{{DATA}}"},
        {"mqtt_topic":"",
        "data":" "}
    ]
}' \
   'http://localhost:13000/tribe/messageBundle?token={{token}}'
   ```